File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if(!root)return 0;
        queue<pair<TreeNode*,int>> q;
        q.push({root,0});
        int ans=1;
        while(!q.empty()){
            int sz=q.size();
            int mn=q.front().second,mx=q.back().second;
            ans=max(ans,mx-mn+1);
            while(sz--){
                auto p=q.front();
                q.pop();
                TreeNode *node=p.first;
                int idx=p.second-mn;
                if(node->left){
                    q.push({node->left,(long long)2*idx+1});
                }
                if(node->right)q.push({node->right,(long long)2*idx+2});
            }
        }
        return ans;
    }
};

File: 0662-maximum-width-of-binary-tree/0662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    
    static class Pair {
        int idx = 0;
        TreeNode node;
        Pair(int idx, TreeNode node) {
            this.idx = idx;
            this.node = node;
        }
    }
    
    public int widthOfBinaryTree(TreeNode root) {
        LinkedList<Pair> q = new LinkedList<>();
        
        q.add(new Pair(0, root));
        int ans = 0;
        while(!q.isEmpty()) {
            int val = q.getLast().idx - q.getFirst().idx + 1;
            ans = Math.max(ans, val);
            int sz = q.size();
            int mn = q.getFirst().idx;
            while(sz-- > 0) {
                Pair p = q.poll();
                TreeNode node = p.node;
                int index = p.idx - mn;
                if(node.left != null) q.add(new Pair(2 * index + 1, node.left));
                if(node.right != null) q.add(new Pair(2 * index + 2, node.right));
                
            }
        }
        
        return ans;
    }
}